core/services/relay/evm/statuschecker/txm_status_checker.go

```diff
@@ -0,0 +1,66 @@
+package statuschecker
+
+import (
+	"context"
+	"fmt"
+	"strings"
+)
+
+//go:generate mockery --quiet --name TransactionStatusChecker --output ../mocks/ --case=underscore
+
+// TODO replace with actual implementation coming from "github.com/smartcontractkit/chainlink-common/pkg/types"
+type TransactionStatus int
+
+const (
+	Unknown TransactionStatus = iota
+	Unconfirmed
+	Finalized
+	Failed
+	Fatal
+)
+
+type TxManager interface {
+	GetTransactionStatus(ctx context.Context, transactionID string) (TransactionStatus, error)
+}
+
+type TransactionStatusChecker interface {
+	CheckMessageStatus(ctx context.Context, msgID string) ([]TransactionStatus, int, error)
+}
+
+type TxmStatusChecker struct {
+	txManager TxManager
+}
+
+type NoOpTxManager struct{}
+
+func (n *NoOpTxManager) GetTransactionStatus() error {
+	return nil
+}
+
+func NewTransactionStatusChecker(txManager TxManager) *TxmStatusChecker {
+	return &TxmStatusChecker{txManager: txManager}
+}
+
+// CheckMessageStatus checks the status of a message by checking the status of all transactions associated with the message ID.
+// It returns a slice of all statuses and the number of transactions found (-1 if none).
+// The key will follow the format: <msgID>-<counter>. TXM will be queried for each key until a NotFound error is returned.
+// The goal is to find all transactions associated with a message ID and snooze messages if they are fatal in the Execution Plugin.
+func (tsc *TxmStatusChecker) CheckMessageStatus(ctx context.Context, msgID string) ([]TransactionStatus, int, error) {
+	var allStatuses []TransactionStatus
+	var counter int
+
+	for {
+		transactionID := fmt.Sprintf("%s-%d", msgID, counter)
+		status, err := tsc.txManager.GetTransactionStatus(ctx, transactionID)
+		if err != nil {
+			if status == Unknown && strings.Contains(err.Error(), fmt.Sprintf("failed to find transaction with IdempotencyKey %s", transactionID)) {
+				break
+			}
+			return nil, counter - 1, err
+		}
+		allStatuses = append(allStatuses, status)
+		counter++
+	}
+
+	return allStatuses, counter - 1, nil
```

Comment:
> Do we need all statuses or just the latest is good enough? Need all right?

Yes we would need all. The main scenario is a message (`mi-0`) submitted at round r0 and only gets resolved at r2. if we fetch the status `mi-2` we won't know what the message is a zko has only `mi-0` resolved

Reply:
> Does it make sense to put a cap (even if its like 1,000) on the # of statuses to return to avoid getting into an infinite loop here?

Good point. And yes, we should probably return an error and consider that something is wrong with the message. So returning an error here will signal the batching logic to snooze it.

Reply:
Rough math:
deltaRound of 30s and 1000 retries so that 1000*30/3600 = 8h.

Say the chain has an outage, we will allow up to 8 hours of outage before dropping the message. It is unlikely to happen but it is leaving us enough room